avocado/plugins/spawners/lxc.py

```diff
@@ -0,0 +1,200 @@
+
+    def initialize(self):
+        section = 'spawner.lxc'
+
+        help_msg = 'Distro for the LXC container'
+        settings.register_option(
+            section=section,
+            key='dist',
+            help_msg=help_msg,
+            default='fedora')
+
+        help_msg = 'Release of the LXC container (depends on the choice of distro)'
+        settings.register_option(
+            section=section,
+            key='release',
+            help_msg=help_msg,
+            default='32')
+
+        help_msg = 'Architecture of the LXC container'
+        settings.register_option(
+            section=section,
+            key='arch',
+            help_msg=help_msg,
+            default='i386')
+
+
+class LXCSpawner(Spawner, SpawnerMixin):
+
+    description = 'LXC (container) based spawner'
+    METHODS = [SpawnMethod.STANDALONE_EXECUTABLE]
+
+    @staticmethod
+    def run_container_cmd(container, command):
+        with LXCStreamsFile() as tmp_out, LXCStreamsFile() as tmp_err:
+            exitcode = container.attach_wait(lxc.attach_run_command, command, stdout=tmp_out, stderr=tmp_err)
+            return exitcode, tmp_out.read(), tmp_err.read()
+
+    @staticmethod
+    async def run_container_cmd_async(container, command):
+        with LXCStreamsFile() as tmp_out, LXCStreamsFile() as tmp_err:
+            pid = container.attach(lxc.attach_run_command, command, stdout=tmp_out, stderr=tmp_err)
+            loop = asyncio.get_event_loop()
+            _, exitcode = await loop.run_in_executor(None, os.waitpid, pid, os.WUNTRACED)
+            return exitcode, tmp_out.read(), tmp_err.read()
+
+    @staticmethod
+    def is_task_alive(runtime_task):
+        if runtime_task.spawner_handle is None:
+            return False
+
+        if not LXC_AVAILABLE:
+            msg = 'LXC python bindings not available on the system'
+            runtime_task.status = msg
+            # we shouldn't reach this point
+            raise RuntimeError("LXC dependency is missing")
+
+        container = lxc.Container(runtime_task.spawner_handle)
+        if not container.defined:
+            return False
+        if not container.running:
+            return False
+
+        status, _, _ = LXCSpawner.run_container_cmd(container, ["pgrep", "avocado-runner"])
+        return status == 0
+
+    async def spawn_task(self, runtime_task):
+        task = runtime_task.task
+        entry_point_cmd = '/root/avocado-runner'
+        entry_point_args = task.get_command_args()
+        entry_point_args.insert(0, "task-run")
+        entry_point_args.insert(0, entry_point_cmd)
+
+        dist = self.config.get('spawner.lxc.distro')
+        release = self.config.get('spawner.lxc.release')
+        arch = self.config.get('spawner.lxc.arch')
+
+        if not LXC_AVAILABLE:
+            msg = 'LXC python bindings not available on the system'
+            runtime_task.status = msg
+            return False
+
+        container_id = self.cid
```

Comment:
I'm missing where it was defined `self.cid`.

Reply:
This is exactly the point: it is not defined anywhere since I was discussing reusing containers in the main thread of this PR. For more details you can check it but the essentially the problem is that we want to reuse containers since some could have 10+ mins setup time that cannot be spent on each individual test. What we do at the moment is that we create the containers as "run slots" (e.g. 3 containers for 3x faster running) and reuse these when a worker picks up a new task for running. This roughly translates to having three processes running at the same time as an upper limit.

I will easily move this to the constructor where it belongs as container name but wanted to first discuss the more global problem with all of you in order to know whether I should just generate a uid here or can actually pass such an argument and control the number of containers from the outside.

Reply:
Turns out using a constructor is not as easy as I thought since the spawner is initiated outside of our runner's scope and an instance accessed via a dispatcher:
```python
spawner = SpawnerDispatcher(job.config, job)[spawner_name].obj
```
Then I started thinking about doing something like this
```diff
diff --git a/avocado/plugins/spawners/lxc.py b/avocado/plugins/spawners/lxc.py
index 09f71b35..200d9950 100644
--- a/avocado/plugins/spawners/lxc.py
+++ b/avocado/plugins/spawners/lxc.py
@@ -122,6 +122,9 @@ class LXCSpawner(Spawner, SpawnerMixin):
         distro = self.config.get('spawner.lxc.distro')
         release = self.config.get('spawner.lxc.release')
         arch = self.config.get('spawner.lxc.arch')
+        # LXC differs from Podman in the fact that we can reuse heavier complete subsystem
+        # and not just lightweight application-centric containers if necessary
+        container_id = self.config.get('spawner.lxc.cid')
 
         if not LXC_AVAILABLE:
             msg = 'LXC python bindings not available on the system'
```
but this doesn't make much sense still since the container ID can be different per test and doesn't sound at all like something that belongs to the overall job config. Essentially what I need is to provide a reusable container ID (container slot) via a config or some task attribute. The way this is done with Podman right now is not reusable since the container ID is returned from a freshly created container each time and thus retrieved within the scope of spawning a task. @richtja @willianrampazzo do you have a good suggestion on what is the best approach here to provide the spawner with some test specific configuration?